networkx/algorithms/approximation/traveling_salesman.py

```diff
@@ -321,6 +333,757 @@ def threshold_tsp(G, weight):
+    nx.Graph
+        A spanning tree using the distribution defined by `gamma`.
+
+    References
+    ----------
+    .. [1] V. Kulkarni, Generating random combinatorial objects, Journal of
+       algorithms, 11 (1990), pp. 185–207
+    """
+
+    def find_node(merged_nodes, n):
+        """
+        We can think of clusters of contracted nodes as having one
+        representative in the graph. Each node which is not in merged_nodes
+        is still its own representative. Since a representative can be later
+        contracted, we need to recursively search though the dict to find
+        the final representative, but only we know it we can use path
+        compression speed the access of the representative for next time.
+
+        Parameters
+        ----------
+        merged_nodes : dict
+            The dict storing the mapping from node to representative
+        n
+            The node whose representative we seek
+
+        Returns
+        -------
+        The representative of the `n`
+        """
+        if n not in merged_nodes:
+            return n
+        else:
+            rep = find_node(merged_nodes, merged_nodes[n])
+            merged_nodes[n] = rep
+            return rep
+
+    def prepare_graph():
+        """
+        For the graph `G`, remove all edges not in the set `V` and then
+        contract all edges in the set `U`.
+
+        Returns
+        -------
+        A copy of `G` which has had all edges not in `V` removed and all edges
+        in `U` contracted.
+        """
+
+        # The result is a MultiGraph version of G so that parallel edges are
+        # allowed during edge contraction
+        result = nx.MultiGraph(incoming_graph_data=G)
+
+        # Remove all edges not in V
+        edges_to_remove = set(result.edges()).difference(V)
+        result.remove_edges_from(edges_to_remove)
+
+        # Contract all edges in U
+        #
+        # Imagine that you have two edges to contract and they share an
+        # endpoint like this:
+        #                        [0] ----- [1] ----- [2]
+        # If we contract (0, 1) first, the contraction function will always
+        # delete the second node it is passed so the resulting graph would be
+        #                             [0] ----- [2]
+        # and edge (1, 2) no longer exists but (0, 2) would need to be contracted
+        # in its place now. That is why I use the below dict as a merge-find
+        # data structure with path compression to track how the nodes are merged.
+        merged_nodes = {}
+
+        for u, v in U:
+            u_rep = find_node(merged_nodes, u)
+            v_rep = find_node(merged_nodes, v)
+            # We cannot contract a node with itself
+            if u_rep == v_rep:
+                continue
+            nx.contracted_nodes(result, u_rep, v_rep, self_loops=False, copy=False)
+            merged_nodes[v_rep] = u_rep
+
+        return merged_nodes, result
+
+    U = set()
+    V = set(G.edges())
```

Comment:
`V` is used in the `prepare_graph` inner function. The first step in that function is to remove all edges which are not in `V`. While we may be removing edges from `U` and `V` at the same time, the two sets will differ since the edges we decide to include are added to `U` earlier in the for loop.